### PR TITLE
Completely redo the substitute-case-callout work

### LINKED ISSUE #564: Improved API for pcre2_set_substitute_case_callout
As requested by Carlo.

He correctly points out that we can't change the API once it goes public. At Microsoft, we're only interested in simple case-transformations (one character to one character), but others may wish to allow multi-character transformations like ß→SS.

So, I guess I should just do the more fancy thing, if I can make time before the 10.45 release.

It will be a somewhat annoying bit of string manipulation to do. Currently, we apply case transformations immediately to characters as they are generated. However, the more powerful API would need to generate all the text first, and _then_ uppercase or lowercase substrings of it. The case transformations need to be able to look ahead and work on whole words at once.

Opening this ticket to acknowledge the feedback, and as a reminder.

### PR DESCRIPTION
**Pull request by NWilson**
Fixes #564

### CODE REVIEW

File: src/pcre2_substitute.c [hunk 1]
```diff
@@ -832,8 +1235,15 @@ do
...
-            CHECKMEMCPY(mark, fraglength);
+            if (forcecase.to_case != PCRE2_SUBSTITUTE_CASE_NONE &&
+                substitute_case_callout == NULL)
+              CHECKCASECPY(mark, fraglength);
+            else
+              CHECKMEMCPY(mark, fraglength);
             }
+
+            // XXX ^^^ This is a behavioural change (to do casing on a MARK)
+            // Ensure we have test coverage, and check it's OK with Philip
```
Comment (NWilson) [inline src/pcre2_substitute.c:1246]: @PhilipHazel I made a behavioural change here. Previously if you did `\U${*MARK}` it would copy the mark without case-transforming it.

I decided this was probably an oversight, rather than a conscious decision, so I fixed it.

**Review by NWilson — COMMENTED** [review]
The new code is:
* More complicated, for the case-callout condition. That's because the case callout stuff is inherently more powerful now.
* Actually simpler, potentially, for the normal no-case-callout condition.
* This PR is supposed to have no behavioural change for users who aren't using the case-callout... except for the one fix to `\U${*MARK}` handling.

File: src/pcre2_substitute.c (hunk 1 above)
Comment (PhilipHazel) [reply src/pcre2_substitute.c:1246]: I assume that it does already case transform /U${1} etc,? I think it certainly ought to, that's the whole point. So yes, I agree.

Comment (NWilson) [reply src/pcre2_substitute.c:1246]: Yes, copying fragments from the replacement string, or from the subject string, were being case-forced. It was just the special case of copying the mark which was using the original casing of the mark.

I just wanted to check, rather than silently "fix" something.

**Comment by NWilson** [comment]
@carenas I did this API change because you requested it. Could you check the new API signature and documentation matches what you wanted?

File: src/pcre2.h.in [hunk 2]
```diff
@@ -496,11 +497,17 @@ For binary compatibility, only add to this list; do not renumber. */
...
-#define PCRE2_SUBSTITUTE_CASE_TITLE 2
+PCRE2_SUBSTITUTE_CASE_LOWER and PCRE2_SUBSTITUTE_CASE_UPPER are passed to the
+callout to indicate that the case of the entire callout input should be
+case-transformed. PCRE2_SUBSTITUTE_CASE_TITLE_FIRST is passed to indicate that
+only the first character or glyph should be transformed to Unicode titlecase,
+and the rest to lowercase. */
+
+#define PCRE2_SUBSTITUTE_CASE_LOWER        1
+#define PCRE2_SUBSTITUTE_CASE_UPPER        2
+#define PCRE2_SUBSTITUTE_CASE_TITLE_FIRST  3
```
Comment (carenas) [inline src/pcre2.h.in:510]: should't these have SIMPLE somewhere on their name, so it is clear which kind of transformation is requested?

assume whoever implements the more complicated version will have to use 3 more defines to indicate they do want a "non simple" transformation.

File: src/pcre2.h.in [hunk 3]
```diff
@@ -673,7 +680,9 @@ PCRE2_EXP_DECL int PCRE2_CALL_CONVENTION \
     int (*)(pcre2_substitute_callout_block *, void *), void *); \
 PCRE2_EXP_DECL int PCRE2_CALL_CONVENTION \
   pcre2_set_substitute_case_callout(pcre2_match_context *, \
-    uint32_t (*)(uint32_t, int, void *), void *); \
+    PCRE2_SIZE (*)(PCRE2_SPTR, PCRE2_SIZE, PCRE2_UCHAR *, PCRE2_SIZE, int, \
```
Comment (carenas) [inline src/pcre2.h.in:683]: should a  size of `PCRE2_ZERO_TERMIMATED` be supported here as a second parameter?

**Review by carenas — APPROVED** [review]
nice work, sorry for the delayed response

File: src/pcre2.h.in (hunk 2 above)
Comment (NWilson) [reply src/pcre2.h.in:510]: These defines are intended for the "non simple" transformation. If all you wanted was simple case transformation, you wouldn't need to set a custom callout.

We're just taking the `\u` and `\l` flags from the replacement string, and telling the callout what was requested by the replacement. The callout can decide to use whatever casing rules it likes to perform that.

File: src/pcre2.h.in (hunk 3 above)
Comment (NWilson) [reply src/pcre2.h.in:683]: There's no NUL. The string which is passed to the callout is not nul-terminated, and we don't require the callout to write out a NUL at the end.

File: src/pcre2_substitute.c [hunk 4]
```diff
@@ -330,18 +742,20 @@ BOOL overflowed = FALSE;
...
 BOOL replacement_only;
 BOOL utf = (code->overall_options & PCRE2_UTF) != 0;
-#ifdef SUPPORT_UNICODE
-BOOL ucp = (code->overall_options & PCRE2_UCP) != 0;
-#endif
 PCRE2_UCHAR temp[6];
 PCRE2_SPTR ptr;
 PCRE2_SPTR repend = NULL;
+// XXX AARGH! All the arithmetic on extra_needed is done without overflow
+// checking - it can totally fail, especially on 32-bit systems
```
Comment (NWilson) [inline src/pcre2_substitute.c:749]: It's not really part of this PR, but I'm tempted to add some overflow checks to the `extra_needed` arithmetic. It's a pre-existing issue. It's just about realistic for a replacement string to have 65,000 copies of "$1", and for $1 to be 65,000 characters long, which would then happily overflow on a 32-bit system.

File: src/pcre2.h.in (hunk 3 above)
Comment (carenas) [reply src/pcre2.h.in:683]: got it, but then is a substitution sting possibly larger than any other string in the library meant to be supported, or should and error be reported in that case, and if so how?

probably PCRE2_NULL might be another interesting error to report if a mechanism to do so is being added here.

Comment (NWilson) [reply src/pcre2.h.in:683]: A string cannot be larger than PCRE2_SIZE; that's supposed to be big enough for the entire address space. PCRE2 supports 32-bit (2GiB) strings in a 32-bit build, and ≈infinite-size strings in a 64-bit build.

The substitute case callout API uses PCRE2_SIZE for the string sizes, so it supports exactly the same string sizes as all the other strings in PCRE2.

The callout will always be given a non-null, non-empty string by PCRE2. It can return `~(PCRE2_SIZE)0` to simply say "something failed". We can't provide more fine-grained errors for the callout to return, because we don't know what might go wrong in the application code.

I don't think we really have a reason to allow the callout to return an error code via an `int*`. The callout can save that state itself anyway.

Comment (carenas) [reply src/pcre2.h.in:684]: the problem is tht we use (the equivalent) of `(size_t)-1` for `PCRE2_ZERO_TERMINATED`, so we need to be ble to answer sensibly if that value (or equivalent) is passed as parameter 2.

returning with an error at least, sure seems like an improvement.

Comment (NWilson) [reply src/pcre2.h.in:684]: Which API exactly are you worried about? The callback API is called by us, back into library client code. We'll never pass `PCRE2_ZERO_TERMINATED` for the length parameters.

Comment (NWilson) [reply src/pcre2.h.in:684]: Thank you! I appreciate your help.
